apps/site/util/download/archive.tsx

```diff
@@ -0,0 +1,183 @@
+import semVer from 'semver';
+
+import type {
+  DownloadDropdownItem,
+  DownloadKind,
+  DownloadArtifact,
+  OperatingSystem,
+  Platform,
+} from '#site/types';
+import type { NodeRelease } from '#site/types/releases';
+import { OS_NOT_SUPPORTING_INSTALLERS, PLATFORMS } from '#site/util/download';
+import { getNodeDownloadUrl } from '#site/util/url';
+
+import { DIST_URL } from '#site/next.constants';
+
+type CompatibleArtifactOptions = {
+  platforms?: Record<OperatingSystem, Array<DownloadDropdownItem<Platform>>>;
+  exclude?: Array<string>;
+  versionWithPrefix: string;
+  kind?: DownloadKind;
+};
+
+/**
+ * Creates a download artifact from platform data
+ */
+const createDownloadArtifact = (
+  os: OperatingSystem,
+  platform: DownloadDropdownItem<Platform>,
+  versionWithPrefix: string,
+  kind: DownloadKind
+): DownloadArtifact => {
+  const url = getNodeDownloadUrl({
+    versionWithPrefix: versionWithPrefix,
+    os: os,
+    platform: platform.value,
+    kind: kind,
+  });
+
+  return {
+    fileName: url.replace(`${DIST_URL}${versionWithPrefix}/`, ''),
+    kind: kind,
+    os: os,
+    architecture: platform.label,
+    url: url,
+    version: versionWithPrefix,
+  };
+};
+
+type CompatiblePlatforms = Array<{
+  os: OperatingSystem;
+  platform: DownloadDropdownItem<Platform>;
+}>;
+
+/**
+ * Filters platforms by compatibility and exclusions
+ */
+const getCompatiblePlatforms = (
+  platforms: Record<OperatingSystem, Array<DownloadDropdownItem<Platform>>>,
+  exclude: Array<string>,
+  versionWithPrefix: string
+): CompatiblePlatforms => {
+  return Object.entries(platforms).flatMap(([os, items]) => {
+    if (exclude.includes(os)) {
+      return [];
+    }
+
+    return items
+      .filter(({ compatibility }) => {
+        // In the download constants file (apps/site/util/download/constants.json),
+        //  if no compatibility is defined as a semver, we treat it as supporting all
+        //  versions by default.
+        if (!compatibility.semver) {
+          return true;
+        }
+
+        return compatibility.semver.every(version =>
+          semVer.satisfies(versionWithPrefix, version)
+        );
+      })
+      .map(platform => ({
+        os: os as OperatingSystem,
```

Comment:
Do you need to add the `os as OperatingSystem` (?) can't you cast the result as `CompatiblePlatforms`